gapidapk/android/app/src/main/java/com/google/android/gapid/Counter.java

```diff
@@ -0,0 +1,175 @@
+     */
+    public static String collect(boolean reset) {
+        StringBuilder sb = new StringBuilder();
+        synchronized (ROOT) {
+            for (Counter thread : ROOT.children.values()) {
+                thread.write(sb, 0);
+            }
+        }
+        if (reset) {
+            ROOT.reset();
+        }
+        return sb.toString();
+    }
+
+    private Counter(String name, Counter parent) {
+        this.name = name;
+        this.parent = parent;
+    }
+
+    private synchronized Counter child(String name) {
+        Counter counter = children.get(name);
+        if (counter == null) {
+            counter = new Counter(name, this);
+            children.put(name, counter);
+        }
+        return counter;
+    }
+
+    private void enter() {
+        CURRENT.set(this);
+    }
+
+    private void exit(float duration) {
+        synchronized (this) {
+            time += duration;
+            count++;
+        }
+        CURRENT.set(parent);
+    }
+
+    private static void indent(StringBuilder sb, int depth) {
+        for (int i = 0; i < depth; i++) {
+            sb.append("    ");
+        }
+    }
+
+    private synchronized void reset() {
+        children.clear();
+        count = 0;
+        time = 0;
+    }
+
+    private synchronized void write(StringBuilder sb, int depth) {
+        if (count > 0) {
+            indent(sb, depth);
+            if (parent.time != 0) {
+                sb.append((int)(100 * time / parent.time)).append("% ");
+            }
+
+            sb.append(name).append(" ")
+                    .append("[total: ").append(time)
+                    .append(" count: ").append(count)
+                    .append(" average: ").append(time / count)
+                    .append("]\n");
+        } else {
+            sb.append(name).append("\n");
+        }
+
+        if (children.size() > 0) {
+            List<Counter> sorted = new ArrayList<>();
+            sorted.addAll(children.values());
+            Collections.sort(sorted, new Comparator<Counter>() {
+                @Override
+                public int compare(Counter a, Counter b) {
+                    if (a.time < b.time) { return 1; }
+                    if (a.time > b.time) { return -1; }
+                    return 0;
+                }
+            });
+            for (Counter child : sorted) {
+                child.write(sb, depth+1);
+            }
+
+            float other = time;
+            for (Counter child : children.values()) {
+                other -= child.time;
+            }
+            if (other > 0) {
+                indent(sb, depth+1);
+                sb.append((int)(100 * other / time)).append("% <other>\n");
+            }
+        }
+    }
+
+    public static class Scope implements AutoCloseable {
+        private final long start;
+        private final Counter counter;
+
+        private Scope(Counter counter) {
+            this.start = System.currentTimeMillis();
```

Comment:
You probably should use `System.nanoTime`, since `System.currentTimeMillis` may not be monotonic and has very loose guarantees.

Reply:
Done.